src/Compilers/CSharp/Test/Semantic/Semantics/StackAllocSpanExpressionsTests.cs

```diff
@@ -0,0 +1,495 @@
+        }
+
+        [Fact]
+        public void ConstStackAllocExpression()
+        {
+            var test = @"
+unsafe public class Test
+{
+    void M()
+    {
+        const int* p = stackalloc int[1];
+    }
+}
+";
+            CreateStandardCompilation(test, options: new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary, allowUnsafe: true)).VerifyDiagnostics(
+                // (6,15): error CS0283: The type 'int*' cannot be declared const
+                //         const int* p = stackalloc int[1];
+                Diagnostic(ErrorCode.ERR_BadConstType, "int*").WithArguments("int*").WithLocation(6, 15));
+        }
+
+        [Fact]
+        public void RefStackAllocAssignment_ValueToRef()
+        {
+            var test = @"
+using System;
+public class Test
+{
+    void M()
+    {
+        ref Span<int> p = stackalloc int[1];
+    }
+}
+";
+            CreateCompilationWithMscorlibAndSpan(test, options: TestOptions.ReleaseDll.WithAllowUnsafe(true)).VerifyDiagnostics(
+                // (7,23): error CS8172: Cannot initialize a by-reference variable with a value
+                //         ref Span<int> p = stackalloc int[1];
+                Diagnostic(ErrorCode.ERR_InitializeByReferenceVariableWithValue, "p = stackalloc int[1]").WithLocation(7, 23),
+                // (7,27): error CS1510: A ref or out value must be an assignable variable
+                //         ref Span<int> p = stackalloc int[1];
+                Diagnostic(ErrorCode.ERR_RefLvalueExpected, "stackalloc int[1]").WithLocation(7, 27));
+        }
+
+        [Fact]
+        public void RefStackAllocAssignment_RefToRef()
+        {
+            var test = @"
+using System;
+public class Test
+{
+    void M()
+    {
+        ref Span<int> p = ref stackalloc int[1];
+    }
+}
+";
+            CreateCompilationWithMscorlibAndSpan(test, options: TestOptions.ReleaseDll.WithAllowUnsafe(true)).VerifyDiagnostics(
+                // (7,31): error CS1510: A ref or out value must be an assignable variable
+                //         ref Span<int> p = ref stackalloc int[1];
+                Diagnostic(ErrorCode.ERR_RefLvalueExpected, "stackalloc int[1]").WithLocation(7, 31));
+        }
+
+        [Fact]
+        public void InvalidPositionForStackAllocSpan()
+        {
+            var test = @"
+using System;
+public class Test
+{
+    void M()
+    {
+        N(stackalloc int[1]);
+    }
+    void N(Span<int> span)
+    {
+    }
+}
+";
+            CreateCompilationWithMscorlibAndSpan(test, options: TestOptions.ReleaseDll.WithAllowUnsafe(true)).VerifyDiagnostics(
+                // (7,9): error CS1525: Invalid expression term 'stackalloc'
+                //         N(stackalloc int[1]);
+                Diagnostic(ErrorCode.ERR_InvalidExprTerm, "N").WithArguments("stackalloc").WithLocation(7, 9));
+        }
```

Comment:
Why is there a syntax error here but not in `OverloadResolution_Fail`?

---
In reply to: [133260037](https://github.com/dotnet/roslyn/pull/21006#discussion_r133260037) [](ancestors = 133260037)